stream.go

```diff
@@ -0,0 +1,126 @@
+package kinesis
+
+import (
+	"fmt"
+	"log"
+	"text/template"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/gliderlabs/logspout/router"
+)
+
+type ErrStreamNotReady struct {
+	s string
+}
+
+func (e *ErrStreamNotReady) Error() string {
+	return fmt.Sprintf("not ready, stream: %s", e.s)
+}
+
+// Stream represents a stream that will send messages to its writer.
+type Stream struct {
+	client     Client
+	name       string
+	tags       *map[string]*string
+	Writer     *writer
+	ready      bool
+	readyWrite chan bool
+	readyTag   chan bool
+}
+
+// NewStream instantiates a new stream.
+func NewStream(name string, tags *map[string]*string, pKeyTmpl *template.Template) *Stream {
+	client := &client{
+		kinesis: kinesis.New(&aws.Config{}),
+	}
+
+	writer := newWriter(
+		newBuffer(pKeyTmpl, name),
+		newFlusher(client.kinesis),
+	)
+
+	s := &Stream{
+		client:     client,
+		name:       name,
+		tags:       tags,
+		Writer:     writer,
+		ready:      false,
+		readyWrite: make(chan bool),
+		readyTag:   make(chan bool),
+	}
+
+	return s
+}
+
+// Start runs the goroutines making calls to create and tag the stream on
+// AWS.
+func (s *Stream) Start() {
+	go s.create()
```

Comment:
You could actually get rid of some of these channels and just make it something as simple as:

```go
func (s *Stream) Write(m *router.Message) error {
        switch {
        case s.err != nil:
                return s.err
        case s.ready:
                s.Writer.Write(m)
                return nil
        default:
                return &ErrStreamNotReady{s: s.name}
        }
}

func (s *Stream) Start() {
        go s.start()
}

func (s *Stream) start() {
        if s.err = s.create(); s.err != nil {
                return
        }

        if s.err = s.tag(); s.err != nil {
                return
        }

        s.ready = true
}
```

It would be nice if we buffered up messages while the stream was in a pending state but probably not for this PR.